In [4]:
from sage.schemes.elliptic_curves.weierstrass_morphism import *
S.<a,b,p,l,u,v,c,d,r,s,t,k,x,y,e>=QQ[]

In [5]:
def Eorig(c,d,e,b):
    A=c^3*d^2*e
    return EllipticCurve(S.fraction_field(),[A,0,A^2*b,0,0])
def F(c,d,e,b):
    return Eorig(c,d,e,b).change_ring(QQ)
def vp(n,p):
    try:
        log(1/Qp(p)(n).abs(),p)
    except ZeroDivisionError:
            return 'inf'
    else:
        return log(1/Qp(p)(n).abs(),p)
def PolyMod(f,n): ##Reduces a polynomial mod n
    R = IntegerModRing(n)
    S.<a,b,p,l,u,v,c,d,r,s,t,x,y,e>=R[]
    return S(f)
def vals(G):
	print("c4")
	print(factor(G.c4()))
	print("c6")
	print(factor(G.c6()))
	print("Disc")
	print(factor(G.discriminant()))
	print('a1')
	print(factor(G.a1()))
	print('a2')
	print(factor(G.a2()))
	print('a3')
	print(factor(G.a3()))
	print('a4')
	print(factor(G.a4()))
	print('a6')
	print(factor(G.a6()))

The curve E(a,b,d) corresponds to $E_{C_{3}}=E_{C_{3}}(a,b,d)$ in the paper where $a=c^{3}d^{2}e$ where $d$ and $e$ are positive relatively prime squarefree integers.

Below, we verify Theorem 3.5. We use the Weierstrass model F1, F2 which is given below:

In [39]:
F1=Eorig(c,d,e,b).change_weierstrass_model(c^2*d,(c^2*d)^2*p^2,0,0)
F2=Eorig(c,d,e,b).change_weierstrass_model(c^2*d,-c^6*d^4*e^2/9,-c^3*d^2*e,c^9*d^6*e^3/27)

**Case 1.** We consider the cases when Nèron type II or III. This is **Case 1 in the proof of Theorem 3.5.**

**Case 1a.** $v_3(a-27b) = 4$.  Note that this implies that $v_3(a) = 3$, $v_3(c)=1$ and $v_3(d)=v_3(e)=0$.

In [10]:
'c4'
print(factor(E(c,d,e,b).c4()))
'c6'
print(factor(E(c,d,e,b).c6()))
'disc'
print(factor(E(c,d,e,b).discriminant()))

c * d^2 * e^3 * (c^3*d^2*e - 24*b)
(-1) * d^2 * e^4 * (c^6*d^4*e^2 - 36*b*c^3*d^2*e + 216*b^2)
b^3 * d^4 * e^8 * (c^3*d^2*e - 27*b)


<span style="color:blue"> In this case, by inspection we get $(v_3(c_4), v_3(c_6), v_3(\Delta))=(1 + v_3(c^3d^2e - 24b),3,4)=(2,3,4)$. Hence, by Papadopoulos, we get the  Nèron type II and the other local data in Case 1 of Theorem 3.5 when $v_3(a-27b) = 4$.

**Case 1b.** $v_3(a) \equiv 0 \mod 3$ and $v_3(a-27b) = 3$.  Note that this implies that $v_3(c) \ge 1$ and $v_3(d)=v_3(e)=0$.

In [28]:
'c4'
print(factor(E(c,d,e,b).c4()))
'c6'
print(factor(E(c,d,e,b).c6()))
'disc'
print(factor(E(c,d,e,b).discriminant()))

print(factor((E(c,d,e,b).c6()/27)^2 - E(c,d,e,b).c4()/3 + 2))
print(PolyMod(64*b^4*d^4*e^8 + 8*b*c*d^2*e^3 + 2,9))

c * d^2 * e^3 * (c^3*d^2*e - 24*b)
(-1) * d^2 * e^4 * (c^6*d^4*e^2 - 36*b*c^3*d^2*e + 216*b^2)
b^3 * d^4 * e^8 * (c^3*d^2*e - 27*b)
(1/729) * (c^12*d^12*e^12 - 72*b*c^9*d^10*e^11 + 1728*b^2*c^6*d^8*e^10 - 15552*b^3*c^3*d^6*e^9 + 46656*b^4*d^4*e^8 - 243*c^4*d^4*e^4 + 5832*b*c*d^2*e^3 + 1458)
b^4*d^4*e^8 + 8*b*c*d^2*e^3 + 2


<span style="color:blue"> In this case, by inspection we get
$$(v_3(c_4), v_3(c_6), v_3(\Delta))=(v_3(c) + 1,3,3)=(\ge2,3,3).$$
Hence by Papadopoulos we get the Nèron type II or III. Next, we verify the supplementary condition:
$$(c_6/27)^2 - c_4/3 + 2=(1/729)(c^{12}d^{12}e^{12} - 72bc^9d^{10}e^{11} + 1728b^2c^6d^8e^{10} - 15552b^3c^3d^6e^9 + 46656b^4d^4e^8 - 243c^4d^4e^4 + 5832bcd^2e^3 + 1458).$$
This gives the following modulo $9$:
$$(c_6/27)^2 - c_4/3 + 2 \equiv 0-0+0-0 +64b^4d^4e^8 - 0 + 8bcd^2e^3 + 2 \equiv bd^2e^3(b^3d^2e^5-c)+2 \mod 9.$$
The other local data in Case 1 of Theorem 3.5 when $v_3(a) \equiv 0 \mod 3$ and $v_3(a-27b) = 3$.

**Case 2.** Case when E_T has Neron type IV and $\rm{IV}^*$. This is **Case 2 in the proof of Theorem 3.5.**

**Subcase 2a.** For any prime $p\neq 3$ with $v_p(a) \equiv  2 \mod 3$. Then $v_p(c)>=0, v_p(d)=1$, and $v_p(e)=0$. We consider the Weierstrass model F1

In [58]:
K=F1
vals(K)

c4
(c^3*d^2*e - 24*b)*c*d^2*e^3
c6
-(c^6*d^4*e^2 - 36*b*c^3*d^2*e + 216*b^2)*d^2*e^4
Disc
(c^3*d^2*e - 27*b)*b^3*d^4*e^8
a1
c*d*e
a2
3*p^2
a3
(c*p^2 + b*e)*d*e
a4
3*p^4
a6
p^6


<span style="color:blue"> By inspection, for the above conditions we get
$$(v_p(a_1),v_p(a_2),v_p(a_3),v_p(a_4),v_p(a_6),v_p(\Delta))=(1+v_p(c),2+v_p(3),1,4+v_p(3),6,4 + v_p(a-27b))=(\ge1,\ge2,1,\ge4,6,4 + v_p(a-27b)).$$
Also, we have $v_p(b_6)=2$ and $v_p(b_8)\ge 3$. So, the Nèron type is IV at $p$.
The other local data in Subcase 2a of Theorem 3.5 now follows from above computation.
</span>

**Subcase 2b.** For any prime $p\neq 3$ with $v_p(a) \equiv 1 \mod 3$. Then $v_p(c)>=0, v_p(d)=0$, and $v_p(e)=1$.  We consider the Weierstrass model F1

In [59]:
K=F1
vals(K)
print('b6')
print(factor(K.b6()))
print('b8')
print(factor(K.b8()))

c4
(c^3*d^2*e - 24*b)*c*d^2*e^3
c6
-(c^6*d^4*e^2 - 36*b*c^3*d^2*e + 216*b^2)*d^2*e^4
Disc
(c^3*d^2*e - 27*b)*b^3*d^4*e^8
a1
c*d*e
a2
3*p^2
a3
(c*p^2 + b*e)*d*e
a4
3*p^4
a6
p^6
b6
c^2*d^2*e^2*p^4 + 2*b*c*d^2*e^3*p^2 + b^2*d^2*e^4 + 4*p^6
b8
(c^2*d^2*e^2*p^4 + 3*b*c*d^2*e^3*p^2 + 3*b^2*d^2*e^4 + 3*p^6)*p^2


<span style="color:blue"> By inspection, for the above conditions we get
$$(v_p(a_1),v_p(a_2),v_p(a_3),v_p(a_4),v_p(a_6),v_p(\Delta))=(1+v_p(c),2+v_p(3),2,4+v_p(3),6,8 + v_p(a-27b))=(\ge1, \ge 2,2, \ge 4,6,8 + v_p(a-27b)).$$
Consequently,
$$P(t)=t^3+a_{2,1}t^2+a_{4,2}t+a_{6,3} \equiv t^3 \mod p$$
and 
$$t^2+a_{3,2}t-a_{6,4} \equiv t(t+a_{3,2}) \mod p.$$
So, by Tate's algorithm the Nèron type is $\rm{IV}^*$ at $p$.
The other local data in Subcase 2b of Theorem 3.5 now follows from above computation.

**Subcase 2c.** $p=3$ and $v_3(a-27b) = 5$. Note that this implies $v_3(a) = 3$ and thus $v_3(c) = 1, v_3(d)=v_3(e)=0$. We consider the Weierstrass model F2.

In [61]:
K=F2
vals(K)
print('b6')
print(factor(K.b6()))
print('b8')
print(factor(K.b8()))

c4
(c^3*d^2*e - 24*b)*c*d^2*e^3
c6
-(c^6*d^4*e^2 - 36*b*c^3*d^2*e + 216*b^2)*d^2*e^4
Disc
(c^3*d^2*e - 27*b)*b^3*d^4*e^8
a1
-c*d*e
a2
-1/3*c^2*d^2*e^2
a3
-1/27*(c^3*d^2*e - 27*b)*d*e^2
a4
-1/27*(c^3*d^2*e - 27*b)*c*d^2*e^3
a6
1/729*(c^3*d^2*e - 27*b)*c^3*d^4*e^5
b6
1/729*(5*c^3*d^2*e - 27*b)*(c^3*d^2*e - 27*b)*d^2*e^4
b8
-1/2187*(2*c^3*d^2*e - 27*b)*(c^3*d^2*e - 27*b)*c^2*d^4*e^6


<span style="color:blue"> 
By inspection, for the above conditions we get
$$(v_3(a_1),v_3(a_2),v_3(a_3),v_3(a_4),v_3(a_6),v_3(\Delta))=(1,1,2,3,2,5).$$
Also, we have $v_3(b_6)=2$ and $v_3(b_8)\ge 3$. So, the Nèron type is IV at $3$.
For Tamagawa number not that
$$t^2+a_{3,1}t-a_{6,2} \equiv t^2 - a_{6,2} = t^2-a(a - 27b)/3^8 \mod 3$$
splits if $a(a - 27b)/3^8\equiv 1 \mod 3$. So, the local data in Subcase 2c of Theorem 3.5 now follows from above computation.

**Case 3.** Suppose $v_3(a-27b)\ge 6$ In particular, $v_3(c) = 1$ and $v_3(d)=v_3(e)=0$. In this case also we consider the Weierstrass model F2. 

In [62]:
K=F2
vals(K)

c4
(c^3*d^2*e - 24*b)*c*d^2*e^3
c6
-(c^6*d^4*e^2 - 36*b*c^3*d^2*e + 216*b^2)*d^2*e^4
Disc
(c^3*d^2*e - 27*b)*b^3*d^4*e^8
a1
-c*d*e
a2
-1/3*c^2*d^2*e^2
a3
-1/27*(c^3*d^2*e - 27*b)*d*e^2
a4
-1/27*(c^3*d^2*e - 27*b)*c*d^2*e^3
a6
1/729*(c^3*d^2*e - 27*b)*c^3*d^4*e^5


<span style="color:blue"> 
By inspection, for the above conditions we get
$$(v_3(c_4),v_3(c_6),v_3(\Delta))=(2,3,v_3(a-27b)).$$
In this case we have the Nèron type $\rm{I}_n^*$ with $n=v_3(a-27b)-6$.
Also, $$(v_3(a_1),v_3(a_2),v_3(a_3),v_3(a_4),v_3(a_6))=(1,1,v_3(a-27b)-3,v_3(a-27b)-2,v_3(a-27b)-3)=(1,1,n+3,n+4,n+3).$$
When $n=0$, we note that $$P(t)=t^3+a_{2,1}t^2+a_{4,2}t+a_{6,3} \equiv t^3-t^2+a(a - 27b)/3^9\mod 3$$
splits if $a(a - 27b)/3^9\equiv 2 \mod 3$.
When $n\ge 1$, we note that
$$t^2+a_{3,\frac{n+3}{2}}t-a_{6,n+3} \equiv t^2-a(a - 27b)3^{-(n+9)} \mod 3 \text{ when } n \text{  is odd},$$
$$a_{2,1}t^2+a_{4,\frac{n+4}{2}}t-a_{6,n+3} \equiv -t^2+a(a - 27b)3^{-(n+9)} \mod 3 \text{ when } n \text{  is even},$$
splits if $a(a - 27b)/3^8\equiv 1 \mod 3$. So, the local data in Subcase 3 of Theorem 3.5 now follows from above computation and Lemma 2.1.